Display error messages for Xero Import page

SzymczakJ opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
This PR adds error messages for Xero Import page.
### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/41843
PROPOSAL: N/A


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. With a Workspace with a Xero connection
2. Turn on offline mode on troubleshoot page
3. Go to Import page
4. Turn on/off some of the import page items
5. Verify that the XeroImportPage looks like this(there's fade on changed items):
<img width="1519" alt="Screenshot 2024-05-13 at 17 02 53" src="https://github.com/Expensify/App/assets/88395093/d4b09245-4cc1-40ce-aa45-532ac542c535">

6. Go into one of the XeroImportPage subpages that got changed and Verify that it looks like this:
<img width="1519" alt="Screenshot 2024-05-13 at 17 03 06" src="https://github.com/Expensify/App/assets/88395093/cbd6a602-8a59-4686-9dbd-92caad6a7374">

7. Turn on simulate failing network requests on troubleshoot page and turn off offline mode
8. Go to Import page and turn on/off some of the import page items
9. Verify that the XeroImportPage looks like this(there's red brick road indicator):
<img width="1518" alt="Screenshot 2024-05-13 at 17 06 19" src="https://github.com/Expensify/App/assets/88395093/a00436fd-0ce9-494c-9bab-bac8c7992a22">

10. Go into one of the XeroImportPage subpages that got changed and Verify that it looks like this:
<img width="1518" alt="Screenshot 2024-05-13 at 17 06 34" src="https://github.com/Expensify/App/assets/88395093/477d3de6-f4ec-4b55-a396-0cccbf63f6ba">



### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
6. Verify a modal appears displaying a preview of that image
--->

Same as tests

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/88395093/9fb5dd0e-4d5a-423d-97fe-640ee29363a0


</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/88395093/b1f39b11-61b7-46de-908b-a9381c2abd2e


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>
https://github.com/Expensify/App/assets/88395093/8d1066f0-b09d-46c6-99ea-98d698018efa

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/Expensify/App/assets/88395093/fc26562d-61ca-4fd9-84d7-d91a0dc6f40c



<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/88395093/fecf8e39-15d4-4b83-9d74-f07d56e71076


<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/88395093/cfe78beb-00e1-4869-9717-dadd98446fad


<!-- add screenshots or videos here -->

</details>

SzymczakJ committed:
display error message on XeroImport

lakchote commented:
@Expensify/design could you please review these error messages if they look good to you?

shawnborton commented:
Before we merge this, should we wait until @narefyev91's [PR here](https://github.com/Expensify/App/pull/41670) is merged? That will give us the updated error style that we want.

lakchote commented:
@trjExpensify this PR will add error messages to the main `Import` page.

@SzymczakJ asked me if we should add error messages to the sub page of the `Import` page also?
It'd look like this below:
![image](https://github.com/Expensify/App/assets/21549771/591692c5-60c2-4881-94f8-34efa395f7f3)

Could we have your opinion on this?

shawnborton commented:
That error style seems way off btw - if we do go with an error on this page, it should stretch to the full width of that view and sit underneath the toggle row:
![image](https://github.com/Expensify/App/assets/2319350/8df4d8dd-25c0-47b9-a1db-d9b447a9d91a)

trjExpensify commented:
>@SzymczakJ asked me if we should add error messages to the sub page of the Import page also?
It'd look like this below:

I think there's typically only one place the error message is and that's at the end of the brick road. So if that's on the Tracking Categories page because of a problem with the import toggle, it should only be there. The "red dots" of the red brick road lead to it.

SzymczakJ commented:
@tjr so you suggest that we shouldn't show error messages on Xero Import Page and instead of that show error messages on Import Page subpages(and show only red dot on Import page). 
It would look something like that:
<img width="374" alt="image" src="https://github.com/Expensify/App/assets/88395093/8ddfe92f-25bc-4d7e-827c-4832840e839c">
<img width="376" alt="image" src="https://github.com/Expensify/App/assets/88395093/bf72458f-c3c6-4e8c-a57d-ebc009589a5e">

cc @trjExpensify @lakchote @shawnborton

trjExpensify commented:
> @trjExpensify so you suggest that we shouldn't show error messages on Xero Import Page and instead of that show error messages on Import Page subpages(and show only red dot on Import page). It would look something like that: <img alt="image" width="374" src="https://private-user-images.githubusercontent.com/88395093/329534992-8ddfe92f-25bc-4d7e-827c-4832840e839c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzMzg1NzMsIm5iZiI6MTcxNTMzODI3MywicGF0aCI6Ii84ODM5NTA5My8zMjk1MzQ5OTItOGRkZmU5MmYtMjViYy00ZDdlLTgyN2MtNDgzMjg0MGU4MzljLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTEwVDEwNTExM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA1N2I3NzczNjIwZGY0NjlhZDJjNjc1NDE2ODc5N2RlMDA3NWEzMTZhNTk3ZThlNTk1MmRkZGQ2YjQ3NDQwMTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6XdTyEkb5AK05r1u6JA6fiYuniZ0zjzJg0Vxn_-pb9o"> <img alt="image" width="376" src="https://private-user-images.githubusercontent.com/88395093/329543861-bf72458f-c3c6-4e8c-a57d-ebc009589a5e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzMzg1NzMsIm5iZiI6MTcxNTMzODI3MywicGF0aCI6Ii84ODM5NTA5My8zMjk1NDM4NjEtYmY3MjQ1OGYtYzNjNi00ZThjLWE1N2QtZWJjMDA5NTg5YTVlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTEwVDEwNTExM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEyYjQ0ZjQzMThmMDYzNzVhZGNjYTIyMzUzMmIyOTBmNzc0MjkyNWE1YzEzNjkwYTlhZTE0ZGVjYmQ4MDg0ZDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nCioDSC1iV63HWhc9FWfSBFcpjPwA01f3X6kl06DCtw">
> 
> cc @trjExpensify @lakchote @shawnborton

Yes exactly, because the end of the red brick road is one page deeper and the end of the road is where the error message appears. 👍

SzymczakJ committed:
refactor ToggleSettingOptionRow

SzymczakJ committed:
add error/offline state to Import subpages

SzymczakJ committed:
delete pending action on XeroImportPage subviews

SzymczakJ committed:
Merge branch 'main' into @szymczak/error-messages-xero-import

SzymczakJ commented:
I've addressed your review comments @Expensify/design, could you please review this? 🙏
Current version of screens is in the test steps
 cc @shawnborton @trjExpensify

SzymczakJ committed:
fix lint

shawnborton commented:
Cool, I think that feels pretty good to me.

filip-solecki commented on src/pages/workspace/accounting/xero/XeroTaxesConfigurationPage.tsx:
```diff
@@ -1,59 +1,42 @@
     const styles = useThemeStyles();
     const policyID = policy?.id ?? '';
-    const {importTaxRates, pendingFields} = policy?.connections?.xero?.config ?? {};
+    const xeroConfig = policy?.connections?.xero?.config;
```
```suggestion
    const xeroConfig = policy?.connections?.xero?.config ?? {};
```
And you don't have to add empty object in each place later

filip-solecki commented on src/pages/workspace/accounting/xero/XeroTaxesConfigurationPage.tsx:
```diff
@@ -1,59 +1,42 @@
             accessVariants={[CONST.POLICY.ACCESS_VARIANTS.ADMIN]}
             policyID={policyID}
             featureName={CONST.POLICY.MORE_FEATURES.ARE_CONNECTIONS_ENABLED}
+            contentContainerStyle={[[styles.pb2, styles.ph5]]}
```
```suggestion
            contentContainerStyle={[styles.pb2, styles.ph5]}
```
Should be enough

filip-solecki commented on src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx:
```diff
@@ -59,29 +61,16 @@ function XeroTrackingCategoryConfigurationPage({policy}: WithPolicyProps) {
+            <ToggleSettingOptionRow
+                title={translate('workspace.accounting.import')}
+                switchAccessibilityLabel={translate('workspace.xero.trackingCategories')}
+                isActive={!!isSwitchOn}
```
There is no need to use `!!` as it is Booelan already

filip-solecki commented on src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx:
```diff
@@ -22,7 +21,10 @@ function XeroTrackingCategoryConfigurationPage({policy}: WithPolicyProps) {
+    const xeroConfig = policy?.connections?.xero?.config;
+    const {importTrackingCategories} = policy?.connections?.xero?.config ?? {};
+
+    const isSwitchOn = Boolean(importTrackingCategories);
```
```suggestion
    const isSwitchOn = !!importTrackingCategories;
```

filip-solecki commented on src/pages/workspace/accounting/xero/XeroTrackingCategoryConfigurationPage.tsx:
```diff
@@ -22,7 +21,10 @@ function XeroTrackingCategoryConfigurationPage({policy}: WithPolicyProps) {
     const styles = useThemeStyles();
     const policyID = policy?.id ?? '';
-    const {importTrackingCategories, pendingFields} = policy?.connections?.xero?.config ?? {};
+    const xeroConfig = policy?.connections?.xero?.config;
```
same suggestion as above

filip-solecki commented on src/pages/workspace/accounting/xero/import/XeroCustomerConfigurationPage.tsx:
```diff
@@ -31,32 +30,22 @@ function XeroCustomerConfigurationPage({policy}: WithPolicyProps) {
+                        wrapperStyle={styles.sectionMenuItemTopDescription}
+                    />
+                }
+                isActive={!!isSwitchOn}
```
same as above

filip-solecki commented on src/pages/workspace/categories/WorkspaceCategoriesSettingsPage.tsx:
```diff
@@ -23,6 +23,7 @@ function WorkspaceCategoriesSettingsPage({policy, route}: WorkspaceCategoriesSet
     const isConnectedToAccounting = Object.keys(policy?.connections ?? {}).length > 0;
     const policyID = route.params.policyID ?? '';
     const [policyCategories] = useOnyx(`${ONYXKEYS.COLLECTION.POLICY_CATEGORIES}${policyID}`);
+    const toggleSubtitle = isConnectedToAccounting ? `${translate('workspace.categories.needCategoryForExportToIntegration')} ${translate('workspace.accounting.qbo')}` : '';
```
is it correct? We check here for any connection(may be xero) and display qbo text

SzymczakJ committed:
Merge branch 'main' into @szymczak/error-messages-xero-import

SzymczakJ committed:
fix PR comments

SzymczakJ committed:
Merge branch 'main' into @szymczak/error-messages-xero-import

SzymczakJ committed:
fix typecheck

SzymczakJ committed:
Merge branch 'main' into @szymczak/error-messages-xero-import

filip-solecki reviewed (approved):
LGTM

SzymczakJ committed:
delete pending action from XeroChartOfAccountsPage

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9093471283) to staging by https://github.com/lakchote in version: 1.4.74-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

Julesssss commented:
FYI, a [bug](https://github.com/Expensify/App/issues/42268) was raised where the RBR error isn't always displayed. As this is a new feature I'm not marking it as a blocker though.

hayata-suenaga commented on src/pages/workspace/categories/WorkspaceCategoriesSettingsPage.tsx:
```diff
@@ -44,7 +53,9 @@ function WorkspaceCategoriesSettingsPage({policy, route}: WorkspaceCategoriesSet
                     <ToggleSettingOptionRow
                         title={translate('workspace.categories.requiresCategory')}
-                        subtitle={isConnectedToAccounting ? `${translate('workspace.categories.needCategoryForExportToIntegration')} ${translate('workspace.accounting.qbo')}` : ''}
+                        titleStyle={styles.textStrong}
```
[The text in the `ToggleSettingOptionRow` shouldn't be bold.](https://expensify.slack.com/archives/C01SKUP7QR0/p1715875595366199) 🙇

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9199894178) to production by https://github.com/chiragsalian in version: 1.4.74-6 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

hayata-suenaga commented:
> tl:dr I think the bold text was there before my PR got merged, but correct me if I'm wrong 🙇

that's fair I think the timing of removing the bold style and this PR probably coincided 😅 Anyway, thank you for your work @SzymczakJ!
